File: pypi/trackfw/commands/identity_wizard.py
"""Shared interactive identity wizard — parity port of
``internal/commands/identity_wizard.go``
(ADR-2026-07-25-wizard-unificado-de-identidade-no-agents-install, D1).

Consumed by both ``trackfw init`` and ``trackfw agents install`` (and the
``update``/``uninstall`` operations that share the same lifecycle command).
A single implementation — two would drift at the first UX tweak.

Flow (D6): preset/custom selection → free-text names (custom mode only) →
nickname → confirmation screen (D3) → persist. Declining the confirmation
loops back to preset selection without persisting anything at all.
"""

from __future__ import annotations

import os
import sys
from typing import Any, Callable

from trackfw import identity
from trackfw.i18n import t as i18n_t

# Fixed list of themed identity presets offered by the wizard, shared
# verbatim with `trackfw init`'s legacy prompt table so neither consumer can
# drift from the other. Mirrors identityPresetOptions() (Go).
IDENTITY_PRESET_LABELS: list[tuple[str, str]] = [
    ("greek", "Panteão grego (Zeus, Apolo, Afrodite...)"),
    ("norse", "Mitologia nórdica (Odin, Thor, Freya...)"),
    ("pioneers", "Pioneiros da computação (Turing, Codd, Knuth...)"),
    ("potter", "Harry Potter (Dumbledore, Snape, Luna...)"),
    ("thrones", "Game of Thrones (Tyrion, Jon, Arya...)"),
    ("tolkien", "Senhor dos Anéis (Gandalf, Aragorn, Arwen...)"),
    ("starwars", "Star Wars (Yoda, Leia, Vader...)"),
    ("chaves", "Chaves (Girafales, Madruga, Chiquinha...)"),
    ("turma", "Turma da Mônica (Franjinha, Cebolinha, Mônica...)"),
    ("egyptian", "Panteão egípcio (Thoth, Ísis, Anúbis...)"),
    ("custom", "Personalizar um a um"),
    ("neutral", "Nomes neutros (padrão)"),
]


def _catalog_item(catalog: dict[str, Any], kind: str, item_id: str) -> dict[str, Any]:
    """Look up a single catalog item by kind+id.

    Every id in identity.known_agent_ids() is expected to exist in the
    agents catalog (identity's known ids and the embedded catalog are meant
    to be kept in lockstep). If one is missing, that is a programming error
    and this fails loudly instead of silently mislabeling a field — mirrors
    the panic in buildCustomIdentityGroup/confirmIdentitySelection (Go).
    """
    for item in catalog[kind]:
        if item["id"] == item_id:
            return item
    raise AssertionError(
        f"identity wizard: agent id {item_id!r} from identity.known_agent_ids() "
        "has no entry in the agents catalog"
    )


def _prompt_choice(title: str, choices: list[tuple[str, str]]) -> str:
    """Simple TTY prompt for a single choice: prints [n] label, reads a
    number. Only ever called when sys.stdin.isatty() — never blocks CI."""
    print(title, file=sys.stderr)
    for index, (_, label) in enumerate(choices, 1):
        print(f"  [{index}] {label}", file=sys.stderr)
    raw = input("> ").strip()
    if not raw:
        return ""
    try:
        idx = int(raw) - 1
        if idx < 0:
            raise IndexError
        return choices[idx][0]
    except (ValueError, IndexError):
        return ""


def _prompt_custom_names(catalog: dict[str, Any], known_ids: list[str], helper_text: str) -> list[str]:
    """Prompt one display name per known agent id, labeled with the agent's
    specialty (item.name + item.description from the catalog) — never the
    raw technical id (ADR D4). Rejects empty/duplicate slugs in a loop."""
    values = [""] * len(known_ids)
    slugs_seen: dict[str, str] = {}
    for index, agent_id in enumerate(known_ids):
        item = _catalog_item(catalog, "agents", agent_id)
        label = f"{item['name']} — {item['description']}"
        while True:
            print(label, file=sys.stderr)
            print(f"  {helper_text}", file=sys.stderr)
            value = input("> ").strip()
            try:
                slug = identity.slugify(value)
            except identity.IdentityError as error:
                print(f"  {error}", file=sys.stderr)
                continue
            if slug in slugs_seen:
                print(
                    f"  slug {slug!r} duplicado com o agente {slugs_seen[slug]!r}",
                    file=sys.stderr,
                )
                continue
            slugs_seen[slug] = agent_id
            values[index] = value
            break
    return values


def _build_identity_config(
    identity_select: str,
    known_ids: list[str],
    custom_display_names: list[str],
    user_nickname: str,
) -> identity.Config:
    """Build the in-memory Config for the chosen wizard path, without
    touching disk. Mirrors buildIdentityConfig (Go) so both the wizard and
    any flag-driven caller build the exact same Config shape from the exact
    same inputs."""
    if identity_select == "custom":
        agents: dict[str, identity.AgentIdentity] = {}
        for index, agent_id in enumerate(known_ids):
            slug = identity.slugify(custom_display_names[index])
            agents[agent_id] = identity.AgentIdentity(
                display_name=custom_display_names[index], slug=slug
            )
        cfg = identity.Config(agents=agents)
    else:
        cfg = identity.preset(identity_select)
    cfg.user_nickname = user_nickname
    return cfg


def _confirm_identity_selection(
    catalog: dict[str, Any], known_ids: list[str], cfg: identity.Config
) -> bool:
    """Render the specialty → display name mapping (ADR D3) and ask for
    confirmation. Returns True iff the user confirmed."""
    nickname_label = i18n_t("identity.wizard.nicknameRowLabel")
    label_width = len(nickname_label)
    labels: list[str] = []
    for agent_id in known_ids:
        item = _catalog_item(catalog, "agents", agent_id)
        labels.append(item["description"])
        label_width = max(label_width, len(item["description"]))

    print()
    print(i18n_t("identity.wizard.confirmHeader"))
    for label, agent_id in zip(labels, known_ids):
        agent = cfg.agents.get(agent_id)
        display_name = agent.display_name if agent else ""
        print(f"  {label:<{label_width}}   →  {display_name}")
    if cfg.user_nickname:
        print(f"  {nickname_label:<{label_width}}   {cfg.user_nickname}")
    print()

    question = i18n_t("identity.wizard.confirmQuestion")
    raw = input(f"? {question} (S/n) ").strip().lower()
    return raw in ("", "s", "sim", "y", "yes")


def run_identity_wizard(catalog: dict[str, Any], home: str) -> tuple[identity.Config, bool]:
    """The single interactive identity wizard consumed by both `trackfw
    init` and `trackfw agents install`
    (ADR-2026-07-25-wizard-unificado-de-identidade-no-agents-install, D1).

    Shows, in order: preset/custom selection, free-text names (custom mode
    only), the user nickname, and a confirmation screen (D3, D6). If the
    user declines the confirmation, the wizard loops back to preset
    selection without persisting anything — it never returns having written
    a partial or rejected config to disk.

    Returns (cfg, persisted). "neutral" (or an empty/blank selection) is the
    "write nothing" path and returns (empty Config, False).
    """
    known_ids = identity.known_agent_ids()

    title_preset = i18n_t("init.prompt.identityPreset")
    title_custom_name = i18n_t("init.prompt.identityCustomName")
    title_nickname = i18n_t("init.prompt.identityNickname")

    while True:
        identity_select = _prompt_choice(title_preset, IDENTITY_PRESET_LABELS)

        if identity_select in ("", "neutral"):
            return identity.Config(), False

        custom_display_names: list[str] = [""] * len(known_ids)
        if identity_select == "custom":
            custom_display_names = _prompt_custom_names(catalog, known_ids, title_custom_name)

        user_nickname = input(f"{title_nickname}: ").strip()

        cfg = _build_identity_config(identity_select, known_ids, custom_display_names, user_nickname)
        identity.validate(cfg, known_ids)

        if not _confirm_identity_selection(catalog, known_ids, cfg):
            # D3: declining returns to preset selection without persisting
            # anything at all.
            continue

        identity.save(home, cfg)
        return cfg, True


# identity_wizard_runner is the indirection callers use to invoke the
# wizard, mirroring Go's `var identityWizardRunner = runIdentityWizard`
# package-var pattern. Tests that need to prove a caller *would* invoke the
# wizard (without actually blocking on real input()) substitute this module
# attribute instead of calling run_identity_wizard directly. Callers MUST
# call through the module attribute (`identity_wizard.identity_wizard_runner(...)`),
# never via a direct `from ... import identity_wizard_runner` — the latter
# captures the function reference at import time and monkeypatching the
# module attribute afterwards would silently not take effect.
identity_wizard_runner: Callable[[dict[str, Any], str], tuple[identity.Config, bool]] = run_identity_wizard


def identity_file_exists(home: str) -> bool:
    """Report whether <home>/.trackfw/identity.json already exists."""
    return os.path.isfile(os.path.join(home, ".trackfw", "identity.json"))


def resolve_identity_preset(value: str) -> tuple[identity.Config | None, bool]:
    """Translate a --identity-preset flag value into a Config to persist.

    "none"/"neutral" mean "do not write anything" — the caller must not
    create ~/.trackfw/identity.json for those values. Any other unknown
    value is always an error, listing the accepted values.
    """
    if value in ("none", "neutral"):
        return None, False
    try:
        cfg = identity.preset(value)
    except identity.IdentityError as error:
        valid = ["none", "neutral"] + identity.preset_names()
        raise identity.IdentityError(
            f"identity-preset invalido {value!r} (validos: {', '.join(valid)})"
        ) from error
    return cfg, True


def apply_identity_preset_flag(preset_value: str, operation: str, home: str) -> None:
    """Resolve and persist --identity-preset for `agents install|update|
    uninstall`, reusing resolve_identity_preset so all callers accept the
    exact same preset names and reject invalid ones with the exact same
    error shape."""
    cfg, should_save = resolve_identity_preset(preset_value)
    if not should_save:
        return
    try:
        identity.validate(cfg, identity.known_agent_ids())
    except identity.IdentityError as error:
        raise identity.IdentityError(f"{operation}: identidade invalida: {error}") from error
    try:
        identity.save(home, cfg)
    except identity.IdentityError as error:
        raise identity.IdentityError(f"{operation}: falha ao gravar identidade: {error}") from error


def should_prompt_identity(kind: str, is_tty: bool, identity_exists: bool, force_flag: bool) -> bool:
    """Decide whether `trackfw agents install` (and the update/uninstall
    operations, which share the same lifecycle command) should show the
    interactive identity wizard. All conditions must hold simultaneously
    (ADR D2):

      1. kind is "agents" — skills have no identity and must never prompt
         (D5).
      2. is_tty — never block a non-interactive run.
      3. either no identity is configured yet, or the user explicitly asked
         to reconfigure it via --identity.

    With an identity already configured and no --identity flag, the wizard
    must NOT reappear — a wizard that shows up on every install becomes an
    incentive to script around it, which quietly kills the feature.
    """
    if kind != "agents":
        return False
    if not is_tty:
        return False
    return (not identity_exists) or force_flag

File: internal/commands/identity_wizard.go
package commands

import (
	"fmt"

	"github.com/charmbracelet/huh"
	"github.com/kgsaran/trackfw/internal/i18n"
	"github.com/kgsaran/trackfw/internal/identity"
	"github.com/kgsaran/trackfw/internal/integrations"
)

// identityWizardRunner is the indirection callers use to invoke the wizard,
// mirroring the integrationsStdinIsTTY package-var pattern already used in
// this package. Tests that need to prove a caller *would* invoke the wizard
// (without actually blocking on a real huh.Form, which requires a terminal)
// substitute this var instead of calling runIdentityWizard directly.
var identityWizardRunner = runIdentityWizard

// runIdentityWizard is the single interactive identity wizard consumed by
// both `trackfw init` and `trackfw agents install`
// (ADR-2026-07-25-wizard-unificado-de-identidade-no-agents-install, D1).
//
// It shows, in order: preset/custom selection, free-text names (custom mode
// only), the user nickname, and a confirmation screen (D3, D6). If the user
// declines the confirmation, the wizard loops back to preset selection
// without persisting anything — it never returns having written a partial
// or rejected config to disk.
//
// The returned bool reports whether cfg was persisted via identity.Save.
// "neutral" (or leaving the hidden-group zero value untouched) is the
// "write nothing" path and returns (zero Config, false, nil). A non-nil
// error means the wizard itself failed (e.g. the huh form was aborted); the
// caller does not need to do any additional cleanup in that case either.
func runIdentityWizard(catalog *integrations.Catalog, home string) (identity.Config, bool, error) {
	knownAgentIDs := identity.KnownAgentIDs()

	titleIdentityPreset := i18n.T("init.prompt.identityPreset")
	titleIdentityCustomName := i18n.T("init.prompt.identityCustomName")
	titleIdentityNickname := i18n.T("init.prompt.identityNickname")

	for {
		var (
			identitySelect string
			userNickname   string
		)
		customDisplayNames := make([]string, len(knownAgentIDs))

		form := huh.NewForm(
			huh.NewGroup(
				huh.NewSelect[string]().
					Title(titleIdentityPreset).
					Options(identityPresetOptions()...).
					Value(&identitySelect),
			),
			buildCustomIdentityGroup(catalog, titleIdentityCustomName, knownAgentIDs, customDisplayNames, func() bool {
				return identitySelect != "custom"
			}),
			huh.NewGroup(
				huh.NewInput().
					Title(titleIdentityNickname).
					Value(&userNickname),
			).WithHideFunc(func() bool {
				return identitySelect == "" || identitySelect == "neutral"
			}),
		)
		if err := form.Run(); err != nil {
			return identity.Config{}, false, err
		}

		cfg, outcome, err := resolveIdentitySelection(home, identitySelect, knownAgentIDs, customDisplayNames, userNickname,
			func(candidate identity.Config) (bool, error) {
				return confirmIdentitySelection(catalog, knownAgentIDs, candidate)
			})
		if err != nil {
			return identity.Config{}, false, err
		}
		switch outcome {
		case identitySkipped:
			return identity.Config{}, false, nil
		case identityDeclined:
			// D3: declining returns to preset selection without persisting
			// anything at all.
			continue
		default:
			return cfg, true, nil
		}
	}
}

// identityOutcome distinguishes the three terminal states of a single wizard
// round. A plain bool cannot: "neutral" (write nothing and finish) and
// "user declined the confirmation" (write nothing and re-ask) are both
// "nothing was saved", but the caller must return in the first case and loop
// in the second.
type identityOutcome int

const (
	// identitySkipped: "neutral" or the hidden-group zero value — the wizard
	// writes nothing at all and the caller is done.
	identitySkipped identityOutcome = iota
	// identityDeclined: the config was built and validated, but the user
	// rejected the confirmation screen; nothing was written.
	identityDeclined
	// identitySaved: the config was validated, confirmed and persisted.
	identitySaved
)

// resolveIdentitySelection is the entire non-interactive half of the wizard:
// the "write nothing" predicate, the Config construction, validation,
// confirmation and persistence — in that exact order. runIdentityWizard is
// its only production caller, so this sequence lives in exactly one place and
// tests exercising it are testing the shipped path.
//
// The confirm callback is the sole interactive step (rendered by
// confirmIdentitySelection in production); it is invoked only after
// identity.Validate has already accepted cfg, and identity.Save runs only
// after confirm returns true. That ordering is the whole point of the
// extraction: a slug collision must abort before anything reaches disk.
func resolveIdentitySelection(home, identitySelect string, knownAgentIDs []string, customDisplayNames []string, userNickname string, confirm func(identity.Config) (bool, error)) (identity.Config, identityOutcome, error) {
	if identitySelect == "" || identitySelect == "neutral" {
		return identity.Config{}, identitySkipped, nil
	}

	cfg, err := buildIdentityConfig(identitySelect, knownAgentIDs, customDisplayNames, userNickname)
	if err != nil {
		return identity.Config{}, identitySkipped, fmt.Errorf("identidade invalida: %w", err)
	}
	if err := identity.Validate(cfg, knownAgentIDs); err != nil {
		return identity.Config{}, identitySkipped, fmt.Errorf("identidade invalida: %w", err)
	}

	confirmed, err := confirm(cfg)
	if err != nil {
		return identity.Config{}, identitySkipped, err
	}
	if !confirmed {
		return identity.Config{}, identityDeclined, nil
	}

	if err := identity.Save(home, cfg); err != nil {
		return identity.Config{}, identitySkipped, fmt.Errorf("falha ao gravar identidade: %w", err)
	}
	return cfg, identitySaved, nil
}

// buildIdentityConfig builds the in-memory Config for the chosen wizard
// path, without touching disk, so the Config shape is built from the wizard
// inputs in exactly one place.
func buildIdentityConfig(identitySelect string, knownAgentIDs []string, customDisplayNames []string, userNickname string) (identity.Config, error) {
	var cfg identity.Config
	if identitySelect == "custom" {
		cfg = identity.Config{Agents: make(map[string]identity.AgentIdentity, len(knownAgentIDs))}
		for index, id := range knownAgentIDs {
			slug, err := identity.Slugify(customDisplayNames[index])
			if err != nil {
				return identity.Config{}, fmt.Errorf("identidade customizada invalida para %q: %w", id, err)
			}
			cfg.Agents[id] = identity.AgentIdentity{DisplayName: customDisplayNames[index], Slug: slug}
		}
	} else {
		preset, err := identity.Preset(identitySelect)
		if err != nil {
			return identity.Config{}, err
		}
		cfg = preset
	}
	cfg.UserNickname = userNickname
	return cfg, nil
}

// identityPresetOptions is the fixed list of themed identity presets offered
// by the wizard, shared verbatim between init and agents install so neither
// consumer can drift from the other.
func identityPresetOptions() []huh.Option[string] {
	return []huh.Option[string]{
		huh.NewOption("Panteão grego (Zeus, Apolo, Afrodite...)", "greek"),
		huh.NewOption("Mitologia nórdica (Odin, Thor, Freya...)", "norse"),
		huh.NewOption("Pioneiros da computação (Turing, Codd, Knuth...)", "pioneers"),
		huh.NewOption("Harry Potter (Dumbledore, Snape, Luna...)", "potter"),
		huh.NewOption("Game of Thrones (Tyrion, Jon, Arya...)", "thrones"),
		huh.NewOption("Senhor dos Anéis (Gandalf, Aragorn, Arwen...)", "tolkien"),
		huh.NewOption("Star Wars (Yoda, Leia, Vader...)", "starwars"),
		huh.NewOption("Chaves (Girafales, Madruga, Chiquinha...)", "chaves"),
		huh.NewOption("Turma da Mônica (Franjinha, Cebolinha, Mônica...)", "turma"),
		huh.NewOption("Panteão egípcio (Thoth, Ísis, Anúbis...)", "egyptian"),
		huh.NewOption("Personalizar um a um", "custom"),
		huh.NewOption("Nomes neutros (padrão)", "neutral"),
	}
}

// buildCustomIdentityGroup builds the huh group for the "Personalizar um a
// um" identity path: one input per known agent id, each labeled with the
// agent's specialty — Item.Name and Item.Description from the catalog — and
// never with the raw technical id (ADR D4). helperText is shown as
// secondary Description text under the specialty label.
//
// Every id in ids is expected to exist in the agents catalog (identity's
// KnownAgentIDs and the embedded catalog are meant to be kept in lockstep);
// if one is missing, that is a programming error and buildCustomIdentityGroup
// fails loudly via panic rather than silently mislabeling a field.
func buildCustomIdentityGroup(catalog *integrations.Catalog, helperText string, ids []string, values []string, hide func() bool) *huh.Group {
	fields := make([]huh.Field, 0, len(ids))
	for index, id := range ids {
		index := index
		id := id
		item, ok := catalog.Item(integrations.KindAgents, id)
		if !ok {
			panic(fmt.Sprintf("identity wizard: agent id %q from identity.KnownAgentIDs() has no entry in the agents catalog", id))
		}
		label := fmt.Sprintf("%s — %s", item.Name, item.Description)
		fields = append(fields, huh.NewInput().
			Title(label).
			Description(helperText).
			Value(&values[index]).
			Validate(func(value string) error {
				slug, err := identity.Slugify(value)
				if err != nil {
					return err
				}
				for j := 0; j < index; j++ {
					if values[j] == "" {
						continue
					}
					otherSlug, err := identity.Slugify(values[j])
					if err != nil {
						continue
					}
					if otherSlug == slug {
						return fmt.Errorf("slug %q duplicado com o agente %q", slug, ids[j])
					}
				}
				return nil
			}))
	}
	return huh.NewGroup(fields...).WithHideFunc(hide)
}

// confirmIdentitySelection renders the specialty → display name mapping
// (ADR D3) and asks for confirmation via huh.NewConfirm. The mapping itself
// is printed with plain fmt (not through a huh field) so its column
// alignment matches the ADR's example layout exactly.
func confirmIdentitySelection(catalog *integrations.Catalog, knownAgentIDs []string, cfg identity.Config) (bool, error) {
	nicknameLabel := i18n.T("identity.wizard.nicknameRowLabel")
	labelWidth := len(nicknameLabel)
	labels := make([]string, len(knownAgentIDs))
	for i, id := range knownAgentIDs {
		item, ok := catalog.Item(integrations.KindAgents, id)
		if !ok {
			panic(fmt.Sprintf("identity wizard: agent id %q from identity.KnownAgentIDs() has no entry in the agents catalog", id))
		}
		labels[i] = item.Description
		if len(labels[i]) > labelWidth {
			labelWidth = len(labels[i])
		}
	}

	fmt.Println()
	fmt.Println(i18n.T("identity.wizard.confirmHeader"))
	for i, id := range knownAgentIDs {
		agent := cfg.Agents[id]
		fmt.Printf("  %-*s   →  %s\n", labelWidth, labels[i], agent.DisplayName)
	}
	if cfg.UserNickname != "" {
		fmt.Printf("  %-*s   %s\n", labelWidth, nicknameLabel, cfg.UserNickname)
	}
	fmt.Println()

	var confirmed bool
	err := huh.NewForm(huh.NewGroup(
		huh.NewConfirm().
			Title(i18n.T("identity.wizard.confirmQuestion")).
			Value(&confirmed),
	)).Run()
	if err != nil {
		return false, err
	}
	return confirmed, nil
}

// shouldPromptIdentity decides whether `trackfw agents install` (and the
// update/uninstall operations, which share executeIntegrationMutation)
// should show the interactive identity wizard. All conditions must hold
// simultaneously (ADR D2):
//
//  1. kind is agents — skills have no identity and must never prompt (D5).
//  2. stdin is a TTY — never block a non-interactive run.
//  3. either no identity is configured yet, or the user explicitly asked to
//     reconfigure it via --identity.
//
// With an identity already configured and no --identity flag, the wizard
// must NOT reappear — a wizard that shows up on every install becomes an
// incentive to script around it, which quietly kills the feature.
func shouldPromptIdentity(kind integrations.ItemKind, isTTY bool, identityExists bool, forceFlag bool) bool {
	if kind != integrations.KindAgents {
		return false
	}
	if !isTTY {
		return false
	}
	return !identityExists || forceFlag
}
